Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -331,6 +345,148 @@ public LibraryDependenciesResult(@Nullable final String[] libraries,
     }
   }
 
+  private void configureImportedLibraries(@NotNull final Collection<String> rootsToAddToLib) {
+    runInEventDispatchThread(new Runnable() {
+      @Override
+      public void run() {
+        ApplicationManager.getApplication().runWriteAction(new Runnable() {
+          public void run() {
+            doConfigureImportedLibraries(rootsToAddToLib);
+          }
+        });
+      }
+    });
+  }
+
+  private static void runInEventDispatchThread(final Runnable runnable) {
+    try {
+      if (SwingUtilities.isEventDispatchThread()) {
+        runnable.run();
+      }
+      else {
+        ApplicationManager.getApplication().invokeAndWait(new Runnable() {
+          public void run() {
+            runnable.run();
+          }
+        }, ModalityState.defaultModalityState());
+      }
+    }
+    catch (Exception e) {
+      LOG.warn(e);
+    }
+  }
+
+  private void doConfigureImportedLibraries(@NotNull final Collection<String> libraries) {
+    for (final Project project : myRootsHandler.getTrackedProjects()) {
+      if (project == null) continue;
+      final DartSdk sdk = DartSdk.getDartSdk(project);
+      if (sdk == null) continue;
+
+      final ProjectFileIndex fileIndex = ProjectRootManager.getInstance(project).getFileIndex();
+      final SortedSet<String> folderPaths = new TreeSet<String>();
+      final Collection<String> rootsToAddToLib = new THashSet<String>();
+
+      for (final String path : libraries) {
+        if (path != null) {
+          folderPaths.add(PathUtil.getParentPath(FileUtil.toSystemIndependentName(path)));
+        }
+      }
+
+      outer:
+      for (final String path : folderPaths) {
+        final VirtualFile vFile = LocalFileSystem.getInstance().findFileByPath(path);
+        if (!path.startsWith(sdk.getHomePath() + "/") && (vFile == null || !fileIndex.isInContent(vFile))) {
+          for (String configuredPath : rootsToAddToLib) {
+            if (path.startsWith(configuredPath + "/")) {
+              continue outer; // folderPaths is sorted so subfolders go after parent folder
+            }
+          }
+          rootsToAddToLib.add(path);
+        }
+      }
+
+
+      final Set<Module> affectedModules = new THashSet<Module>();
+      final Module[] modules = ModuleManager.getInstance(project).getModules();
+      for (final Module module : modules) {
+        if (DartSdkGlobalLibUtil.isDartSdkGlobalLibAttached(module, sdk.getGlobalLibName())) {
+          for (final VirtualFile contentRoot : ModuleRootManager.getInstance(module).getContentRoots()) {
+            // if there is a pubspec, skip this contentRoot
+            if (contentRoot.findChild(PubspecYamlUtil.PUBSPEC_YAML) != null) continue;
```

Comment:
Do you think it will be better to check for pubspec.yaml file presence not only in the content root, but anywhere at any nested level in any of module content roots? 
`FilenameIndex.getVirtualFilesByName(module.getProject(), PUBSPEC_YAML, module.getModuleContentScope())`
If no pubspec file in module at all then we think that this is google internal module. Is there any better way to detect google internal project?